modules/openxr/extensions/spatial_entities/openxr_spatial_entities.cpp

```diff
@@ -0,0 +1,505 @@
+	ClassDB::bind_method(D_METHOD("get_parent", "index"), &OpenXRSpatialComponentParentList::get_parent);
+}
+
+void OpenXRSpatialComponentParentList::set_capacity(uint32_t p_capacity) {
+	parent_data.resize(p_capacity);
+
+	parent_list.parentCount = uint32_t(parent_data.size());
+	parent_list.parents = parent_data.ptrw();
+}
+
+XrSpatialComponentTypeEXT OpenXRSpatialComponentParentList::get_component_type() const {
+	return XR_SPATIAL_COMPONENT_TYPE_PARENT_EXT;
+}
+
+void *OpenXRSpatialComponentParentList::get_structure_data(void *p_next) {
+	parent_list.next = p_next;
+	return &parent_list;
+}
+
+RID OpenXRSpatialComponentParentList::get_parent(int64_t p_index) const {
+	ERR_FAIL_INDEX_V(p_index, parent_data.size(), RID());
+
+	OpenXRSpatialEntityExtension *se_extension = OpenXRSpatialEntityExtension::get_singleton();
+	ERR_FAIL_NULL_V(se_extension, RID());
+
+	return se_extension->find_spatial_entity(parent_data[p_index]);
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Spatial component mesh2d list
+
+void OpenXRSpatialComponentMesh2DList::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_transform", "index"), &OpenXRSpatialComponentMesh2DList::get_transform);
+	ClassDB::bind_method(D_METHOD("get_vertices", "snapshot", "index"), &OpenXRSpatialComponentMesh2DList::get_vertices);
+	ClassDB::bind_method(D_METHOD("get_indices", "snapshot", "index"), &OpenXRSpatialComponentMesh2DList::get_indices);
+}
+
+void OpenXRSpatialComponentMesh2DList::set_capacity(uint32_t p_capacity) {
+	mesh2d_data.resize(p_capacity);
+
+	mesh2d_list.meshCount = uint32_t(mesh2d_data.size());
+	mesh2d_list.meshes = mesh2d_data.ptrw();
+}
+
+XrSpatialComponentTypeEXT OpenXRSpatialComponentMesh2DList::get_component_type() const {
+	return XR_SPATIAL_COMPONENT_TYPE_MESH_2D_EXT;
+}
+
+void *OpenXRSpatialComponentMesh2DList::get_structure_data(void *p_next) {
+	mesh2d_list.next = p_next;
+	return &mesh2d_list;
+}
+
+Transform3D OpenXRSpatialComponentMesh2DList::get_transform(int64_t p_index) const {
+	ERR_FAIL_INDEX_V(p_index, mesh2d_data.size(), Transform3D());
+
+	OpenXRAPI *openxr_api = OpenXRAPI::get_singleton();
+	ERR_FAIL_NULL_V(openxr_api, Transform3D());
+
+	return openxr_api->transform_from_pose(mesh2d_data[p_index].origin);
+}
+
+PackedVector2Array OpenXRSpatialComponentMesh2DList::get_vertices(RID p_snapshot, int64_t p_index) const {
+	ERR_FAIL_INDEX_V(p_index, mesh2d_data.size(), PackedVector2Array());
+
+	const XrSpatialBufferEXT &buffer = mesh2d_data[p_index].vertexBuffer;
+	if (buffer.bufferId == XR_NULL_SPATIAL_BUFFER_ID_EXT) {
+		// We don't have data (yet).
+		return PackedVector2Array();
+	}
+
+	ERR_FAIL_COND_V(buffer.bufferType != XR_SPATIAL_BUFFER_TYPE_VECTOR2F_EXT, PackedVector2Array());
+
+	OpenXRSpatialEntityExtension *se_extension = OpenXRSpatialEntityExtension::get_singleton();
+	ERR_FAIL_NULL_V(se_extension, PackedVector2Array());
+
+	return se_extension->get_vector2_buffer(p_snapshot, buffer.bufferId);
```

Comment:
Just as a reminder to myself, I'll probably do this in a follow up.
Currently we retrieve this data and check in `set_mesh_data` if it has changed, which we should keep as a fallback.

However feedback from the OpenXR WG suggests that we can add an extra check here. We should cache `buffer.bufferId` in our calling logic, if it is unchanged from the last time, the data is unchanged and we don't need to call this getter.

When new data is available, we should always have a new buffer id.

This applies to our other buffer getter functions as well.